cloudweatherreport/model.py

```diff
@@ -0,0 +1,755 @@
+        return self
+
+    @classmethod
+    def from_json(cls, value):
+        """
+        Deserialize an instance of this model from a JSON string.
+
+        The data will be validated during deserialization.
+        """
+        return cls.from_dict(json.loads(value))
+
+    @classmethod
+    def from_yaml(cls, value):
+        """
+        Deserialize an instance of this model from a YAML string.
+
+        The data will be validated during deserialization.
+        """
+        return cls.from_dict(yaml.safe_load(value))
+
+
+# ********  Cloud Weather Report models
+
+class BenchmarkPlan(BaseModel):
+    fields = {
+        'unit': basestring,
+        'action': basestring,
+        'params': dict,
+    }
+
+
+class TestPlan(BaseModel):
+    fields = {
+        'benchmarks': list([BenchmarkPlan]),
+        'bundle': basestring,
+        'bundle_file': basestring,
+        'bundle_name': basestring,
+        'tests': list([basestring]),
+    }
+
+    @classmethod
+    def load_plans(cls, filename):
+        """
+        Load a test_plan.yaml file.
+
+        Returns a list of TestPlan instances.
+        """
+        with open(filename) as fp:
+            return cls.from_dict_or_list(yaml.safe_load(fp))
+
+    @classmethod
+    def from_dict_or_list(cls, data):
+        """
+        Deserialize a TestPlan instance from either a single
+        Python dictionary or a list of dictionaries.
+
+        The data will be validated during deserialization.
+        """
+        if not isinstance(data, list):
+            data = [data]
+        return list(map(TestPlan.from_dict, data))
+
+    @classmethod
+    def from_dict(cls, data):
+        """
+        Deserialize a TestPlan instance from a Python dictionary.
+
+        The data will be validated during deserialization.
+
+        Benchmark plans are normalized into a more usable structure.
+        """
+        benchmarks = data.pop('benchmark', {})
+        plan = super(TestPlan, cls).from_dict(data)
+        if benchmarks:
+            for unit, action_info in benchmarks.items():
+                if isinstance(action_info, basestring):
+                    action_info = {action_info: {}}
+                for action, params in action_info.items():
+                    plan.benchmarks.append(BenchmarkPlan(
+                        unit=unit,
+                        action=action,
+                        params=params,
+                    ))
+        return plan
+
+    def report_filename(self, test_id):
+        return Report(
+            test_id=test_id,
+            bundle=BundleInfo(name=self.bundle),
+        ).filename_json
+
+
+class BundleInfo(BaseModel):
+    fields = {
+        'machines': None,
+        'name': basestring,
+        'relations': None,
+        'services': None,
+    }
+
+
+class BenchmarkProviderResult(BaseModel):
+    fields = {
+        'provider': basestring,
+        'value': float,
```

Comment:
I can't think of a better name than "value", but the naming choice had me reading the next function a few times before I figured out what it meant to do. (It looked like it was trying to convert all values to floats at first, rather than just converting to a float if this was the "value" field).